src/stack/Stack.h

```diff
@@ -0,0 +1,295 @@
+    CHIP_ERROR Shutdown() { return CHIP_NO_ERROR; }
+
+    PersistentStorageDelegate * Get() { return nullptr; }
+};
+
+/* ========================== BleLayer Configuration ========================== */
+struct BleLayerConfiguration
+{
+};
+template <typename T>
+struct IsBleLayerConfiguration : std::is_base_of<BleLayerConfiguration, T>
+{
+};
+
+class DefaultBleLayer : BleLayerConfiguration
+{
+public:
+    CHIP_ERROR Init(const StackParameters & parameters) { return CHIP_NO_ERROR; }
+    CHIP_ERROR Shutdown() { return CHIP_NO_ERROR; }
+
+    Ble::BleLayer * Get() { return nullptr; }
+};
+
+/* ========================== Transport Configuration ========================== */
+struct TransportConfiguration
+{
+};
+template <typename T>
+struct IsTransportConfiguration : std::is_base_of<TransportConfiguration, T>
+{
+};
+
+class DefaultTransport : TransportConfiguration
+{
+public:
+    using transport = TransportMgr<
+#if INET_CONFIG_ENABLE_IPV4
+        Transport::UDP, /* IPv4 */
+#endif
+        Transport::UDP>;
+
+    CHIP_ERROR Init(const StackParameters & parameters, Inet::InetLayer & inetLayer, Ble::BleLayer * bleLayer)
+    {
+        return mTransportManager.Init(
+#if INET_CONFIG_ENABLE_IPV4
+            Transport::UdpListenParameters(&inetLayer)
+                .SetAddressType(Inet::kIPAddressType_IPv4)
+                .SetListenPort(parameters.GetListenPort()),
+#endif
+            Transport::UdpListenParameters(&inetLayer)
+                .SetAddressType(Inet::kIPAddressType_IPv6)
+                .SetListenPort(parameters.GetListenPort()));
+    }
+
+    CHIP_ERROR Shutdown()
+    {
+        mTransportManager.Close();
+        return CHIP_NO_ERROR;
+    }
+
+    TransportMgrBase & Get() { return mTransportManager; }
+
+private:
+    transport mTransportManager;
+};
+
+/* ========================== Stack ========================== */
+
+/**
+ * @brief
+ *   Chip Stack is an interface to manage all chip modules for applications.
+ *
+ *   This class should not be used inside CHIP SDK.
+ */
+template <typename... Configurations>
+class Stack
+{
+private:
+    using SystemLayerConfig =
+        typename first_if_any_or_default<IsSystemLayerConfiguration, DefaultSystemLayer, Configurations...>::type;
+    using InetLayerConfig = typename first_if_any_or_default<IsInetLayerConfiguration, DefaultInetLayer, Configurations...>::type;
+    using StorageConfig   = typename first_if_any_or_default<IsStorageConfiguration, DefaultStorage, Configurations...>::type;
+    using BleLayerConfig  = typename first_if_any_or_default<IsBleLayerConfiguration, DefaultBleLayer, Configurations...>::type;
+    using TransportConfig = typename first_if_any_or_default<IsTransportConfiguration, DefaultTransport, Configurations...>::type;
+
+public:
+    Stack(NodeId localDeviceId) : mLocalDeviceId(localDeviceId) {}
+    virtual ~Stack() {}
+
+    // TODO(#6931): CHIP initialization should be posted to chip thread, in order to avoid racing problems
```

Comment:
Is this true? Isn't this code supposed to be called before the chip thread is started?

Reply:
IMHO, the initialization will split to 3 parts:

 1. Initialize the platform: run in main thread, do hardware initialization
 2. Create the chip thread
 3. Initialize the chip components: run in chip thread

I'll work on it in the following PRs. Use issue #6931 to track the progress.